lib/cli/src/commands/run/wasi.rs

```diff
@@ -225,21 +249,161 @@ impl Wasi {
         Ok((wasi_env, instance))
     }
 
+    // Runs the Wasi process
+    pub fn run(run: RunProperties, store: Store) -> Result<i32> {
+        let tasks = run.ctx.data(&store).tasks().clone();
+
+        // The return value is passed synchronously and will block until the result is returned
+        // this is because the main thread can go into a deep sleep and exit the dedicated thread
+        let (tx, rx) = std::sync::mpsc::channel();
+
+        // We run it in a blocking thread as the WASM function may otherwise hold
+        // up the IO operations
+        tasks.task_dedicated(Box::new(move || {
+            Self::run_with_deep_sleep(run, store, tx, None);
+        }))?;
+        rx.recv()
+            .expect("main thread terminated without a result, this normally means a panic occurred within the main thread")
+    }
+
+    // Runs the Wasi process (asynchronously)
+    pub fn run_with_deep_sleep(
+        run: RunProperties,
+        mut store: Store,
+        tx: Sender<Result<i32>>,
+        rewind_state: Option<(RewindState, Result<(), Errno>)>,
+    ) {
+        // If we need to rewind then do so
+        let ctx = run.ctx;
+        if let Some((mut rewind_state, trigger_res)) = rewind_state {
+            if rewind_state.is_64bit {
+                if let Err(exit_code) = rewind_state
+                    .rewinding_finish::<Memory64>(ctx.env.clone().into_mut(&mut store), trigger_res)
+                {
+                    tx.send(Ok(exit_code.raw())).ok();
+                    return;
+                }
+                let res = rewind::<Memory64>(
```

Comment:
This is way too much logic to expose/put into the CLI.

This should be handled by a helper function in the `wasi` crate.

Reply:
I've reverted all this code back to synchronous only, given its isolated to the CLI only anyway its not that big a deal to revert it.